src/python/pants/util/memo.py

```diff
@@ -24,21 +24,37 @@ def equal_args(*args, **kwargs):
   return key
 
 
+class InstanceKey(object):
+  """An equality wrapper for an arbitrary object instance.
+
+  This wrapper leverages `id` and `is` for fast `__hash__` and `__eq__` but both of these rely on
+  the object in question not being gc'd since both `id` and `is` rely on the instance address which
+  can be recycled; so we retain a strong reference to the instance to ensure no recycling can occur.
+  """
+
+  def __init__(self, instance):
+    self._instance = instance
+    self._hash = id(instance)
+
+  def __hash__(self):
+    return self._hash
+
+  def __eq__(self, other):
+    if self._instance is other:
```

Comment:
This is FUD - we'll only get InstanceKey others since this is all private-use, but I'll leave for follow-up instead of burning more coal.